linalg/operator.hpp

```diff
@@ -384,62 +391,83 @@ class TimeDependentOperator : public Operator
    { eval_mode = new_eval_mode; }
 
    /** @brief Perform the action of the explicit part of the operator, G:
-       @a y = G(@a x, t) where t is the current time.
+       @a v = G(@a u, t) where t is the current time.
 
        Presently, this method is used by some PETSc ODE solvers, for more
        details, see the PETSc Manual. */
-   virtual void ExplicitMult(const Vector &x, Vector &y) const;
+   virtual void ExplicitMult(const Vector &u, Vector &v) const;
 
    /** @brief Perform the action of the implicit part of the operator, F:
-       @a y = F(@a x, @a k, t) where t is the current time.
+       @a v = F(@a u, @a k, t) where t is the current time.
 
        Presently, this method is used by some PETSc ODE solvers, for more
        details, see the PETSc Manual.*/
-   virtual void ImplicitMult(const Vector &x, const Vector &k, Vector &y) const;
+   virtual void ImplicitMult(const Vector &u, const Vector &k, Vector &v) const;
 
-   /** @brief Perform the action of the operator: @a y = k = f(@a x, t), where
-       k solves the algebraic equation F(@a x, k, t) = G(@a x, t) and t is the
-       current time. */
-   virtual void Mult(const Vector &x, Vector &y) const;
-
-   /** @brief Solve the equation: @a k = f(@a x + @a dt @a k, t), for the
-       unknown @a k at the current time t.
-
-       For general F and G, the equation for @a k becomes:
-       F(@a x + @a dt @a k, @a k, t) = G(@a x + @a dt @a k, t).
-
-       The input vector @a x corresponds to time index (or cycle) n, while the
-       currently set time, #t, and the result vector @a k correspond to time
-       index n+1. The time step @a dt corresponds to the time interval between
-       cycles n and n+1.
-
-       This method allows for the abstract implementation of some time
-       integration methods, including diagonal implicit Runge-Kutta (DIRK)
-       methods and the backward Euler method in particular.
+   /** @brief Perform the action of the explicit part of the operator, G:
+       @a v = G(@a u, t) where t is the current time.
+
+       For solving an ordinary differential equation of the form
+       \f$ M \frac{dy}{dt} = g(y,t) \f$, recall F and G are defined as one of
+       the following:
+       1. F(u,k,t) = k and G(u,t) = inv(M) g(u,t)
+       2. F(u,k,t) = M k and G(u,t) = g(u,t)
+
+       This function then computes one of the following:
+       1. @a v = inv(M) g(@a u, t)
+       2. @a v = g(@a u, t) */
+   virtual void Mult(const Vector &u, Vector &v) const override;
+
+   /** @brief Solve for unknown @a k at current time t that satisfies
+       F(@a u + @a gamma @a k, @a k, t) = G(@a u + @a gamma @a k, t).
+
```

Comment:
I see you removed some of the existing documentation for `ImplicitSolve()` -- I personally think it is best to keep the previous text and just add more explanations.

Reply:
I believe that existing documentation for `ImplicitSolve` was causing a lot of confusion (see #3443 ).  I personally found
```
The input vector @a x corresponds to time index (or cycle) n, while the
currently set time, #t, and the result vector @a k correspond to time
index n+1. The time step @a dt corresponds to the time interval between
cycles n and n+1.
```
to be particularly confusing for a DIRK method, where 
1. the stage solves should evaluate the RHS of the ode at t_n + c_i \Delta t (not time index n +1) 
2. the input vector x is the solution a time index n plus a combination of other stage values
3. the dt should correspond to a_ii \Delta t... which is not always the same as c_i \Delta t 

This was the motivation for my replacing of the text above with
```
A DIRK integrator can use @a k from this function, with @a u set to
\f$ y(t) + \Delta t \sum_{j=1}^{i-1} a_{ij} k_j \f$ and @a gamma set to
\f$ a_{ii} \Delta t \f$, for \f$ k_i \f$.
```

Reply:
The paragraph you found confusing can be fixed by prefixing it by "For, example, in the case of backward Euler ..." -- it is a simple case that is easy to understand and gives some idea how the method works, so I think it is worth keeping with the fix.

Also, the other deleted paragraph expressing the functionality in terms of $f(u,t)$ is also worth having.

Reply:
Now that the method parameter is called `gamma`, the paragraph about backward Euler should mention that in this case `gamma` is `dt` -- "the time interval between cycles n and n+1".

Reply:
> The paragraph you found confusing can be fixed by prefixing it by "For, example, in the case of backward Euler ..." -- it is a simple case that is easy to understand and gives some idea how the method works, so I think it is worth keeping with the fix.

I will add back the case of backward Euler.

>Also, the other deleted paragraph expressing the functionality in terms of $f(u,t)$ is also worth having.

I believe this is captured, with the new notation, in line 430:

```
In either case, this function solves for @a k in M @a k = g(@ a + @a gamma @k, t).
```
What do you think?

Reply:
I'm looking at the latest version, I'll comment there.